op-node/rollup/driver/state_tob_test.go

```diff
@@ -0,0 +1,253 @@
+func (d TestDummyDerivationPipeline) SafeL2Head() eth.L2BlockRef                     { return d.l2SafeHead }
+func (d TestDummyDerivationPipeline) UnsafeL2Head() eth.L2BlockRef                   { return d.l2Finalized }
+
+type TestDummyL1OriginSelector struct {
+	retval eth.L1BlockRef
+}
+
+func (l TestDummyL1OriginSelector) FindL1Origin(ctx context.Context, l1Head eth.L1BlockRef, l2Head eth.L2BlockRef) (eth.L1BlockRef, error) {
+	return l.retval, nil
+}
+
+// fixes "undefined: derive.Progress"
+// func (d TestDummyDerivationPipeline) Progress() derive.Progress {
+// 	return derive.Progress{
+// 		Origin: eth.L1BlockRef{},
+// 		Closed: false,
+// 	}
+// }
+
+// TestRejectCreateBlockBadTimestamp tests that a block creation with invalid timestamps will be caught.
+// This does not test:
+// - The findL1Origin call (it is hardcoded to be the head)
+// - The outputInterface used to create a new block from a given payload.
+// - The DerivationPipeline setting unsafe head (a mock provider is used to pretend to set it)
+// - Metrics (only mocked enough to let the method proceed)
+// - Publishing (network is set to nil so publishing won't occur)
+func TestRejectCreateBlockBadTimestamp(t *testing.T) {
+	// Create our random provider
+	rng := rand.New(rand.NewSource(rand.Int63()))
+
+	// Create our context for methods to execute under
+	ctx := context.Background()
+
+	// Create our fake L1/L2 heads and link them accordingly
+	l1HeadRef := testutils.RandomBlockRef(rng)
+	l2HeadRef := testutils.RandomL2BlockRef(rng)
+	l2l1OriginBlock := l1HeadRef
+	l2HeadRef.L1Origin = l2l1OriginBlock.ID()
+
+	// Create a rollup config
+	cfg := rollup.Config{
+		BlockTime: uint64(60),
+		Genesis: rollup.Genesis{
+			L1:     l1HeadRef.ID(),
+			L2:     l2HeadRef.ID(),
+			L2Time: 0x7000, // dummy value
+		},
+	}
+
+	// Patch our timestamp so we fail
+	l2HeadRef.Time = l2l1OriginBlock.Time - (cfg.BlockTime * 2)
+
+	// Create our outputter
+	outputProvider := TestDummyOutputImpl{willError: false}
+
+	// Create our state
+	s := Driver{
+		l1State: &L1State{
+			l1Head:  l1HeadRef,
+			log:     log.New(),
+			metrics: &metrics.Metrics{TransactionsSequencedTotal: prometheus.NewCounter(prometheus.CounterOpts{})},
+		},
+		log:              log.New(),
+		l1OriginSelector: TestDummyL1OriginSelector{retval: l1HeadRef},
+		config:           &cfg,
+		sequencer:        outputProvider,
+		derivation:       TestDummyDerivationPipeline{},
+		metrics:          &metrics.Metrics{TransactionsSequencedTotal: prometheus.NewCounter(prometheus.CounterOpts{})},
+	}
+
+	// Create a new block
+	// - L2Head's L1Origin, its timestamp should be greater than L1 genesis.
+	// - L2Head timestamp + BlockTime should be greater than or equal to the L1 Time.
+	err := s.createNewL2Block(ctx)
+
+	// Verify the L1Origin's timestamp is greater than L1 genesis in our config.
+	if l2l1OriginBlock.Number < s.config.Genesis.L1.Number {
+		assert.Nil(t, err)
+		return
```

Comment:
`require.NoError(t, err, "message why we don't expect an error here")` would be more idiomatic. The `require` essentially asserts and fails the test there, instead of having to return to stop execution.